### ISSUE TITLE
Death to SSLv3

### ISSUE THREAD

**Issue by DomT4** [Homebrew/homebrew#34634]
Alright, I'm punting this into public discussion. If there aren't any unworkable rejections of the idea, my desire is to press ahead with this and close it out before December 11th.

I'm proposing, primarily, amongst other things, that we remove SSLv3 support from OpenSSL. This is in part driven by POODLE, which has thankfully hastened the demise of SSLv3, but I don't think anyone is going to deny that SSLv3 has been weak at best for years and years now. SSLv3 was 'birthed' in 1996. It was replaced by TLSv1.0 in 1999. 15 years after the introduction of TLSv1.0 we're still in a world where SSLv3 is a thing, sigh. But POODLE has finally nudged the world into some kind of coordinated action, and I think Homebrew should be part of that.

I'll bullet point things from this point on to avoid making anyone read walls of text and scaring everyone off:

* GnuTLS will stop supporting SSLv3 [from 3.4.0](http://nmav.gnutls.org/2014/10/what-about-poodle.html).
* cURL now has SSLv3 [disabled by default](http://daniel.haxx.se/blog/2014/10/17/curl-is-no-poodle/).
* PolarSSL has been [advising people](https://polarssl.org/tech-updates/security-advisories/polarssl-security-advisory-2014-03-poodle-attack-on-ssl-v3) to disable SSLv3. Homebrew disabled SSLv3 in PolarSSL a couple of months back, in [this commit](https://github.com/Homebrew/homebrew/commit/73ca769de13b72682d0dcfe7d54f31f96fd5e5ca). So far, no hiccups.
* Mozilla's latest stable release of Firefox [completely axed](https://blog.mozilla.org/security/2014/10/14/the-poodle-attack-and-the-end-of-ssl-3-0/) SSLv3, and attempting to connect to a site that only talks SSLv3 is automatically blocked for safety.
* Google is [axing](https://groups.google.com/a/chromium.org/forum/#!topic/security-dev/Vnhy9aKM_l4) SSLv3 from [Chromium](https://code.google.com/p/chromium/issues/detail?id=418848&q=sslv3&colspec=ID%20Pri%20M%20Iteration%20ReleaseBlock%20Cr%20Status%20Owner%20Summary%20OS%20Modified), Canary and [Google Chrome](http://www.zdnet.com/google-advances-ssl-with-new-chrome-versions-7000035966/) (>40). Chrome 39 [killed the](https://twitter.com/agl__/status/534756432889929729) SSLv3 fallback mechanism.
* Python have released [two patches](http://bugs.python.org/issue22638) that disable SSLv3 in Python 2.7 & [Python 3.4](https://hg.python.org/cpython/rev/653dfb1240d5). Those patches would be adopted as part of this. Given those patches are in the trunk, It looks inevitable that we will see them at the next stable release.
* Ruby recently [killed SSLv2 and SSLv3](https://bugs.ruby-lang.org/issues/9424). I believe those changes are in the latest stable Ruby release, and have been [backported](https://bugs.ruby-lang.org/issues/9640), although I will check this.
* Cloudflare have some [interesting stats](https://blog.cloudflare.com/sslv3-support-disabled-by-default-due-to-vulnerability/) on SSLv3 usage, which is an important consideration here. Less than 0.7% of *all* SSL/TLS traffic that strolls across Cloudflare's network uses SSLv3.
* Debian Jessie and above [killed SSLv3](http://metadata.ftp-master.debian.org/changelogs//main/o/openssl/openssl_1.0.1j-1_changelog) in early October. So far, nothing has cropped up that has forced Debian to reconsider this. Homebrew has a history of following Debian Jessie in terms of what releases we use and the configuration we adopt, and I feel confident that two months after Debian made the move we can follow them into disabling SSLv3. I am a strong proponent of Debian's move towards better security in recent years, ignoring that slightly embarrassing massive goof-up from 2006-2008. [Debian recently shot down the idea of bringing SSLv3 support back](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=765565).
*  LibreSSL's latest stable release killed off SSLv3 by default, although there is [some confusion](http://comments.gmane.org/gmane.os.openbsd.tech/39606) about that.

I made the argument at the time we disabled SSLv2, and I'll make it again now:

For better or worse, OpenSSL themselves have largely maintained a position over the years that it is up to distro managers and individual users how they configure OpenSSL. OpenSSL ships with all the ciphers you could possibly imagine (*almost*), and it's up to you to decide which ones you want to support or not. So it is down to package managers to decide whether to act or not, rather than waiting on OpenSSL to do so. For better or worse, OpenSSL has decided for now not to make that choice for package managers and individuals.

The OpenSSL trunk has started experimenting with the idea of Security Levels, where mandating Security Level 3 at build time would disable x,y,z ciphers, but that has not yet landed in stable and probably won't land until at least 1.0.2 stable.

The only OpenSSL release which is really going to get smacked about by any SSLv3 change is the 0.9.8 branch, which Apple currently uses, but Apple's OpenSSL support is depreciated. As of December 2015 OpenSSL 0.9.8 will cease to be supported entirely, and the last OpenSSL version which supports nothing newer than TLS1.0 will cease to be shipped by everyone, distros included, *hopefully*. 

Sadly, there isn't a huge amount of point in disabling SSLv3 in OpenSSL 0.9.8 because that leaves it with TLS1.0 and that's all. I'd rather Homebrew ditch supporting OpenSSL 0.9.8 entirely, but I'm content to leave it in ``` Homebrew/Versions ``` until the December 2015 expiration date.

So, We'd do the same dance we did when we killed SSLv2, which is essentially:
* Change OpenSSL, Revision.
* Revision every dependent.

Yes, Making users bump everything is tedious, but yes, I believe it is worthwhile. The only formulae we had significant issues with last time were:
* Python 2.7
* Python 3.4
* Ruby

I seem to recall that [all three were bottle issues](https://github.com/Homebrew/homebrew/issues/31851), although I could be incorrect in that recollection. Now that SSLv3 removal has landed in Ruby stable, and been backported, and there are patches for the two Python versions available that will be in the next stable releases, I feel like those formulae should have a smoother ride this time. 

We've had one request to turn [SSLv2 back on](https://github.com/Homebrew/homebrew/issues/33066) in the time since we disabled it.

I expect SSLv3 to be *slightly* more troublesome, but in reality SSLv3 usage is tiny, and servers *should not* be using SSLv3 any more, and the number of servers out there that can't talk at least TLS1.0 is *tiny*.

Other things I'd like to deal with whilst we're here:

* Disabling IDEA, with ``` no-idea ```. Debian and [Ubuntu](https://launchpadlibrarian.net/178068934/buildlog_ubuntu-utopic-arm64.openssl_1.0.1f-1ubuntu6_UPLOADING.txt.gz) do this. IDEA has a well known weak-key problem, but isn't broken *as such* yet. Still, I feel we'd not lose anything following the Linux distros' lead on killing ``` IDEA ``` support. Apple's native OpenSSL also kills [IDEA](https://opensource.apple.com/source/OpenSSL098/OpenSSL098-27/src/crypto/opensslconf.h). IDEA is enabled by default, so this would be a change, albeit a minor one.
* Specify ``` no-mdc2 ``` and ``` no-rc5 ```. These are already turned off by default, but making that more explicit won't do any harm, and follows the Linux distro's lead.

The most recent OpenSSL release finally properly killed off SSLv3. It turned out that although the option of doing ``` ./configure no-ssl3 ``` had been there for a while it wasn't completely working. As of the latest stable release, which Homebrew is using, that issue has been fixed and ``` no-ssl3 ``` works as expected. There's a slightly odd patch on the current [OpenSSL git](https://git.openssl.org/gitweb/?p=openssl.git;a=commit;h=961d2ddb4b48e0e857a704b0cc6b475d63372419) which I want to ask upstream about first, and pull into our OpenSSL formula if necessary.

Alright, that's me. I open the floor to comments :wink:.

**Comment by azet** [comment]
lol. https://github.com/Homebrew/homebrew/pull/30504

**Comment by DomT4** [comment]
^ We're three+ months down the line, post-POODLE *(i.e. in a world where there's actual mainstream movement against SSLv3 finally)*, and OpenSSL have made the necessary changes upstream to properly support this change. As we discussed at the time of your PR, there was a lot of work still to be done by upstream to actually *properly* turn off the SSLv3 tap. OpenSSL have done that work now, and the new 1.0.2 branch is the ideal place for us to merge a change like this in.

By all means, feel free to contribute to the discussion, but there were valid reasons for rejecting the idea months ago and valid reasons for accepting the idea now - One of the biggest being the relatively new mainstream rejection of the protocol.

**Comment by azet** [comment]
@DomT4 
Sorry, but no. There were no valid reasons to reject the idea. I do this for web hosting as well. I do consulting in that area. I agree that the `no-ssl3` flag caused errors in OpenSSL. **These** have been fixed by now (as well as other very good improvements, eg. constant time NIST ECC curves et cetera). That does not mean that if the `openssl(1)` command fails there's reason enough to stop work on reasonable security for clients/customers! 

Basically everything below TLS 1.2 is **fucked** (yea, I like swearwords, comes from 10yrs of doing operations). Don't take my word for it, as a reference, here's Google's main TLS expert saying the same thing: https://www.imperialviolet.org/2014/12/08/poodleagain.html (last paragraph).

To be clear: I did not want -- at all -- to offend anybody. But things turned ugly pretty quickly. I seriously apologize for that. Your code of conduct is something that I very much agree with. I'm probably just a very impatient person when it comes to end-user security and me being questioned on TLS (in)security statements.

Thanks <3,
Aaron

**Comment by mikemcquaid** [comment]
Not interested in doing this again @azet; I'm just going to edit out anything that's not conducive to sensible discussion. I'm well within my rights to delete my own tweets, edit and delete comments on this repository and will continue to do so. You can express yourself on your personal blog, Twitter, whatever. If you'd rather delete your comments than have me edit them: that's up to you.

**Comment by DomT4** [comment]
With respect, we've had this discussion before. Yes, It is *well known* TLS1.2 is the last *really* secure protocol - Yes, Google and AGL have acknowledged and said this much. However, TLS1.2 adoption *sucks* server-side. Adam L has said (*repeatedly*) that TLS1.2 adoption sucks - It currently sits at an all-time high of 52% global adoption across SSL/TLS available sites. That is a whole 48% of servers that *will not* talk TLS1.2.

Look, I would love to disable everything but TLS1.2 right now, would love to. I'd love to get the world suddenly using TLS1.2 and PFS everywhere, but it's not happening overnight, and no package manager is going to break SSL/TLS for over 50% of their users' connections. Nobody. TLS1.0 has 99.7% adoption support; that is, right now, the comfortable limit.

At some point, you have to draw a line between ultimate security and users being able to do *anything*, and this is the line that Debian is adopting, and various other Linux distros are working towards adopting, and Homebrew is not going to suddenly pole-vault over those distros and break absolutely everything for end users.

**Comment by azet** [comment]
WTF @mikemcquaid is censoring my comments (which were actually useful [and logged!])?! what kind of project is this?!

**Comment by azet** [comment]
Just to be sure that the other post isn't going to be deleted, let me be verbose and post this in a second message: if you like the way apple handles security issues and deals with customers, you're on the right path @mikemcquaid. you're not acting like a democratic open source project but a monolithic dictatorship  of the few. Meaning: the few devs. working on this project full time. I've /never/ seen something like this before in 15 years.

**Comment by DomT4** [comment]
> yes, I see you're working for them - that doesn't give you the right to censor other peoples comments, especially if they are not offensive at all

This literally isn't why Mike has the ability to edit comments. He's one of the maintainers here - They can all edit or delete comments as desired. His workplace has nothing to do with his ability to moderate comments here. Any maintainer of any project on Github has the same power.

Can we not turn this useful thread into some kind of grudge match please Azet? I *would like* to keep the thread open for further user comment.

**Comment by azet** [comment]
@DomT4 

I do agree. This is absolutely useless to any aim of a security goal for this project (I still like the idea of Homebrew).

The issue remains; I know that /any/ project maintainer can edit comments. That doesn't mean you'll have to edit everything out that doesn't fall short of worshipping the project you're involved with. I do run my own projects and mailing lists. I've never done this, and would never do this. This is clearly censorship by any definition in a textbook on political science or the english language!

**Comment by mikemcquaid** [comment]
For anyone following/linked to this thread: @azet has been pointed to our code of conduct multiple times and seems to be unwilling or unable to behave politely. Homebrew welcomes criticism but most of all we welcome well-implemented improvements that don't break more than they fix. There are relatively few maintainers, none of us have ever received any money for working on Homebrew and all of us work on it in our free time. As a result we're happy to help people contribute to the project but we have no interest in moderating a forum for mudslinging. As a result, comments may be edited or deleted for moderation purposes. You do not have the right to say whatever you want on our issue tracker. If you have something you really need to tell the world may I suggest you do so on Twitter or your personal blog.

Any further comments by @azet in this thread will be immediately deleted. He is well within his rights to delete the previous comments I've edited if he feels they no longer represent his position.

**Comment by azet** [comment]
@mikemcquaid 

I have voiced public concern about the security of this particular project. You have censored my messages in this thread -- even though they have not been disruptive just because I do not agree with you. As a committed and active antifascist; I can tell that your twitter description falls short of explaining that your political views might censor other peoples opinions if they do not align! I'm sorry but this is **fucking** fascism (yes, this time I'm swearing)!

I had *constructive* criticism on **parts** of your project. I'm really sorry that you're taking this so personally. But this is a very serious issue. If you think opensource projects are not meant to have free speech you're completely misguided. I value the work you and -- especially -- other maintainers of Homebrew are doing. This has **nothing** to do with my messages in this thread!

All and everything is going to be web-scraped from now-on to insure consistency of my messages. To take this further; if you're censoring messages in simple pull requests, how do we know you're not subverting the security of the entire project on behalf of a Nation Sate Adversary? This is  the most dubious thing I've ever witnessed in an "open source" project.

**Comment by mikemcquaid** [comment]
@azet Please leave me alone.

**Comment by mikemcquaid** [comment]
@azet continues to send me abusive tweets despite being blocked:
![screen shot 2015-01-10 at 18 29 40](https://cloud.githubusercontent.com/assets/125011/5692679/bf53372c-98f6-11e4-817a-61808f84cd95.png)

@azet threatens me by @ mentioning my employer:
![screen shot 2015-01-10 at 18 30 09](https://cloud.githubusercontent.com/assets/125011/5692680/c0a89f4a-98f6-11e4-8adf-dfb61ef1ae9f.png)

**Comment by mikemcquaid** [comment]
Now reporting on my physical whereabouts and seems to hint that he's going to harass me

**Comment by azet** [comment]
@mikemcquaid 

I did not say I'm going to harass you. Is this about you or your projects security? I'd rather improve the security of Homebrew! I'd really like to talk about this over a beer in usual FOSDEM fashion. I can see that you're not interested in speaking to me, still I *really* don't like to be censored. And that's exactly what you're doing. This is just the wrong approach and might cause Homebrew users to worry about the openness of this project. It seems a bit like a dictatorship.

I know I haven't been posting a technical message in my last replies. Just because you're trying to make this about **you** and not the actual technical challenges at hand! Your last couple of messages have been simple slander. And yes. I stand by calling you an asshole. I'm not going to delete this as you're doing with your tweets. Because you are a fascist!

**Comment by mikemcquaid** [comment]
This thread has been locked temporarily as @azet continues to try and post in it.

**Comment by mikemcquaid** [comment]
Unlocked the issue as @azet has apologised privately and we've sorted a bunch of stuff out. Thanks @azet.

**Comment by DomT4** [comment]
Thanks guys. Appreciate the unlocking, and glad things have been resolved to a happier ground :+1:

**Comment by DomT4** [comment]
Per the open PR at #35060, closing this. No point having the discussion in two places I guess. In around 10 days, we get to make a decision as to whether to go ahead with this idea or postpone it indefinitely.
